guides/ruby-on-rails-openid-connect-integration/index.md

```diff
@@ -0,0 +1,192 @@
+
+```rb
+Rails.application.routes.draw do
+  get 'welcome/index'
+  root 'welcome#index'
+end
+```
+
+
+### Outline of Authentication Flow
+
+Authenticating End Users takes two steps.
+
+1. Redirect a user to DID.app with an authentication request from your application.
+2. Fetch user information associated with the code returned in the authentication response.
+
+### Redirect user to authenticate with DID.app
+
+Start the authentication process by redirecting a user to the authorization endpoint.
+The redirect must include the authentication request parameters.
+
+```erb
+<form action="https://auth.did.app/oidc/authorize" method="get">
+  <input name="client_id" value="<%= ENV["CLIENT_ID"] %>" type="hidden" />
+  <input name="redirect_uri" value="<%= callback_url %>" type="hidden" />
+
+  <button type="submit">Sign in</button>
+</form>
+```
+
+The `CLIENT_ID` is a value provided to you by DID.app that we will generate later.  
+The `callback_url` refers to a controller action that we will implement in the next section.
+
+### Create a session controller to handle authentication callback
+
+Once a user has authenticated, with DID.app, they will be redirected to the url that was specified in the authentication request.
+We need to create a route to handle this.
+
+First add the route into `routes.rb`.
+
+```rb
+Rails.application.routes.draw do
+  # Other routes
+  # ...
+
+  get '/session/callback', to: 'session#callback', as: 'callback'
+end
+```
+
+Create a new file for a session controller.
+Add the following code to implement the callback route.
+
+```rb
+require 'faraday'
+require 'json'
+
+class SessionController < ApplicationController
+  def callback
+    response = Faraday.post(
+      "https://auth.did.app/oidc/token",
+      client_id: ENV["CLIENT_ID"],
+      client_secret: ENV["CLIENT_SECRET"],
+      code: params["code"]
+    )
+
+    session[:current_user_id] = data["userinfo"]["sub"]
+    redirect_to root_path
+  end
+end
+```
+
+The redirect url back to your app will include a query parameter containing a one-time code.
+To finish authenticating a user this code needs to be exchanged for a token and user information.
+
+From this user information we extract the `sub` parameter, short for `subject`, this parameter is a stable identifier for the user.
+A verified email address is also returned as part of the user information.
+
+This request will require your `CLIENT_ID` and `CLIENT_SECRET`.
+It is good practice to keep these values out of your source code and so we are using environment variables to configure them.
+
+### Display Sign Out Button to Authenticated User
+
+Once a user has authenticated we no longer want to show them a sign in button.
+Probably the most helpful thing to show instead would be a sign out button.
+
+```erb
+<% if session[:current_user_id] %>
+  <%= link_to 'Sign out', sign_out_path %>
+<% else %>
+  <!-- Sign in form from earlier -->
```

Comment:
Any reason why we wouldn't just put the form in again here?  If there is a reason can we link to an anchor on the page?  Not a blocker just would be nice.

Reply:
Just though this was easier for focusing on what is new for this part of the guide